File: examples/player-python/ViewActions.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the feather-tk project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl

import weakref

class Actions:
    """
    This class provides view actions.
    """
    def __init__(self, context, app, mainWindow):

        self._mainWindowWeak = weakref.ref(mainWindow)
        self.actions = {}
        self.actions["Frame"] = ftk.Action(
            "Frame",
            "ViewFrame",
            ftk.KeyShortcut(ftk.Key.Backspace),
            checkedCallback = self._frameCallback)
        self.actions["Frame"].tooltip = "Toggle whether to automatically frame the view."

        self.actions["ZoomReset"] = ftk.Action(
            "Zoom Reset",
            "ViewZoomReset",
            ftk.KeyShortcut(ftk.Key._0),
            callback = lambda: self._mainWindowWeak().getViewport().viewZoomReset())
        self.actions["ZoomReset"].tooltip = "Reset the view zoom."

        self.actions["ZoomIn"] = ftk.Action(
            "Zoom In",
            "ViewZoomIn",
            ftk.KeyShortcut(ftk.Key.Equals),
            callback=  lambda: self._mainWindowWeak().getViewport().viewZoomIn())
        self.actions["ZoomIn"].tooltip = "Zoom the view in."

        self.actions["ZoomOut"] = ftk.Action(
            "Zoom Out",
            "ViewZoomOut",
            ftk.KeyShortcut(ftk.Key.Minus),
            callback = lambda: self._mainWindowWeak().getViewport().viewZoomOut())
        self.actions["ZoomOut"].tooltip = "Zoom the view out."

        selfWeak = weakref.ref(self)
        self._playerObserver = tl.PlayerObserver(
            app.getDocumentModel().observePlayer(),
            lambda player: selfWeak()._playerUpdate(player))

        self._frameObserver = ftk.BoolObserver(
            mainWindow.getViewport().observeFrameView,
            lambda value: selfWeak()._frameUpdate(value))

    def _frameCallback(self, value):
        self._mainWindowWeak().getViewport().frameView = value

    def _frameUpdate(self, value):
        self.actions["Frame"].checked = value

    def _playerUpdate(self, player):
        self.actions["Frame"].enabled = player != None
        self.actions["ZoomReset"].enabled = player != None
        self.actions["ZoomIn"].enabled = player != None
        self.actions["ZoomOut"].enabled = player != None

File: lib/tlRender/PlayApp/ViewActions.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the tlRender project.

#include "ViewActions.h"

#include "App.h"
#include "FilesModel.h"

#include <tlRender/UI/Viewport.h>

namespace tl
{
    namespace play
    {
        void ViewActions::_init(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<ui::Viewport>& viewport)
        {
            auto viewportWeak = std::weak_ptr<ui::Viewport>(viewport);
            _actions["Frame"] = ftk::Action::create(
                "Frame",
                "ViewFrame",
                [viewportWeak](bool value)
                {
                    if (auto viewport = viewportWeak.lock())
                    {
                        viewport->setFrameView(value);
                    }
                });
            _actions["Frame"]->setTooltip("Toggle whether to automatically frame the view.");

            _actions["ZoomReset"] = ftk::Action::create(
                "Zoom Reset",
                "ViewZoomReset",
                [viewportWeak]
                {
                    if (auto viewport = viewportWeak.lock())
                    {
                        viewport->viewZoomReset();
                    }
                });
            _actions["ZoomReset"]->setTooltip("Reset the view zoom.");

            _actions["ZoomIn"] = ftk::Action::create(
                "Zoom In",
                "ViewZoomIn",
                [viewportWeak]
                {
                    if (auto viewport = viewportWeak.lock())
                    {
                        viewport->viewZoomIn();
                    }
                });
            _actions["ZoomIn"]->setTooltip("Zoom the view in.");

            _actions["ZoomOut"] = ftk::Action::create(
                "Zoom Out",
                "ViewZoomOut",
                [viewportWeak]
                {
                    if (auto viewport = viewportWeak.lock())
                    {
                        viewport->viewZoomOut();
                    }
                });
            _actions["ZoomOut"]->setTooltip("Zoom the view out.");

            _frameObserver = ftk::Observer<bool>::create(
                viewport->observeFrameView(),
                [this](bool value)
                {
                    _actions["Frame"]->setChecked(value);
                });

            _playerObserver = ftk::Observer<std::shared_ptr<Player> >::create(
                app->getFilesModel()->observePlayer(),
                [this](const std::shared_ptr<Player>& value)
                {
                    _actions["Frame"]->setEnabled(value.get());
                    _actions["ZoomReset"]->setEnabled(value.get());
                    _actions["ZoomIn"]->setEnabled(value.get());
                    _actions["ZoomOut"]->setEnabled(value.get());
                });
        }

        ViewActions::~ViewActions()
        {}

        std::shared_ptr<ViewActions> ViewActions::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<ui::Viewport>& viewport)
        {
            auto out = std::shared_ptr<ViewActions>(new ViewActions);
            out->_init(context, app, viewport);
            return out;
        }

        const std::map<std::string, std::shared_ptr<ftk::Action> >& ViewActions::getActions() const
        {
            return _actions;
        }
    }
}
